### DISCUSSION TITLE
.NET Assembly Trimming

### DISCUSSION THREAD

**Discussion by dahall** [dahall/Vanara#341] [General]
Issue #261 prompted the addition of Trimming ([see MS docs](https://docs.microsoft.com/en-us/dotnet/core/deploying/trimming/prepare-libraries-for-trimming)) to Vanara. COM and Trimming are not very friendly. This discussion is a place for the community to share ideas and research around how to enhance Trimming support in Vanara.

**Comment by 0x5bfa** [comment]
Why not use GeneratedComInterface(Class)Attribute for com interop? It's what Hez said, the source generators for com interfaces/classes that internally utilize ComWrappers.

**Reply by 0x5bfa** [reply]
@dahall any thoughts? Or you could rely on CsWin32 for the sake of easier maintenance.

**Reply by dahall** [reply]
CsWin32 is an incredible project, but it focuses on `unsafe` implementations as does `GeneratedComInterface(Class)Attribute`. It would be a significant departure from Vanara where I have tried to avoid `unsafe` code exposed to the consumer. I am actually in the process of learning and testing my own generator based on the meta information that CsWin32 uses. That meta information is lacking some of the detail I need to make consistent use of enumerated values, memory handling, safe handles, overloads, and much more.

**Reply by 0x5bfa** [reply]
Yeah CsWin32 cannot use SG-oriented runtime features like LibraryImport and GeneratedComInterface/Class since itself is a SG. I understand you.

> I am actually in the process of learning and testing my own generator based on the meta information that CsWin32 uses.

Does that mean you're gonna generate managed code with these SG-oriented features? That sounds so so cool.

**Reply by dahall** [reply]
> Does that mean you're gonna generate managed code with these SG-oriented features? That sounds so so cool.

I'm trying, but it is proving to be very complicated.

**Reply by catmanjan** [reply]
I'm not sure when, but since cswin32 now supports GeneratedComInterface

**Reply by 0x5bfa** [reply]
It's because that uses build task instead of being a Roslyn source generator.
